File: tests/test_return_value_policy_pack.cpp
#include <pybind11/functional.h>
#include <pybind11/stl.h>

#include "pybind11_tests.h"

#include <array>
#include <map>
#include <optional>
#include <set>
#include <string>
#include <utility>
#include <variant>
#include <vector>

namespace {

using PairString = std::pair<std::string, std::string>;

PairString return_pair_string() { return PairString({"", ""}); }

using NestedPairString = std::pair<PairString, PairString>;

NestedPairString return_nested_pair_string() {
    return NestedPairString(return_pair_string(), return_pair_string());
}

using MapString = std::map<std::string, std::string>;

MapString return_map_string() { return MapString({return_pair_string()}); }

using MapPairString = std::map<PairString, PairString>;

MapPairString return_map_pair_string() { return MapPairString({return_nested_pair_string()}); }

using SetPairString = std::set<PairString>;

SetPairString return_set_pair_string() { return SetPairString({return_pair_string()}); }

using VectorPairString = std::vector<PairString>;

VectorPairString return_vector_pair_string() { return VectorPairString({return_pair_string()}); }

using ArrayPairString = std::array<PairString, 1>;

ArrayPairString return_array_pair_string() { return ArrayPairString({return_pair_string()}); }

using OptionalPairString = std::optional<PairString>;

OptionalPairString return_optional_pair_string() {
    return OptionalPairString(return_pair_string());
}

using VariantPairString = std::variant<PairString>;

VariantPairString return_variant_pair_string() { return VariantPairString(return_pair_string()); }

std::string call_callback_pass_pair_string(const std::function<std::string(PairString)> &cb) {
    auto p = return_pair_string();
    return cb(p);
}

} // namespace

TEST_SUBMODULE(return_value_policy_pack, m) {
    auto rvpc = py::return_value_policy::_clif_automatic;
    auto rvpb = py::return_value_policy::_return_as_bytes;

    m.def("return_tuple_str_str", []() { return return_pair_string(); });
    m.def(
        "return_tuple_bytes_bytes", []() { return return_pair_string(); }, rvpb);
    m.def(
        "return_tuple_str_bytes",
        []() { return return_pair_string(); },
        py::return_value_policy_pack({rvpc, rvpb}));
    m.def(
        "return_tuple_bytes_str",
        []() { return return_pair_string(); },
        py::return_value_policy_pack({rvpb, rvpc}));

    m.def("return_nested_tuple_str", []() { return return_nested_pair_string(); });
    m.def(
        "return_nested_tuple_bytes", []() { return return_nested_pair_string(); }, rvpb);
    m.def(
        "return_nested_tuple_str_bytes_bytes_str",
        []() { return return_nested_pair_string(); },
        py::return_value_policy_pack({{rvpc, rvpb}, {rvpb, rvpc}}));
    m.def(
        "return_nested_tuple_bytes_str_str_bytes",
        []() { return return_nested_pair_string(); },
        py::return_value_policy_pack({{rvpb, rvpc}, {rvpc, rvpb}}));

    m.def("return_dict_str_str", []() { return return_map_string(); });
    m.def(
        "return_dict_bytes_bytes", []() { return return_map_string(); }, rvpb);
    m.def(
        "return_dict_str_bytes",
        []() { return return_map_string(); },
        py::return_value_policy_pack({rvpc, rvpb}));
    m.def(
        "return_dict_bytes_str",
        []() { return return_map_string(); },
        py::return_value_policy_pack({rvpb, rvpc}));

    m.def(
        "return_map_sbbs",
        []() { return return_map_pair_string(); },
        py::return_value_policy_pack({{rvpc, rvpb}, {rvpb, rvpc}}));
    m.def(
        "return_map_bssb",
        []() { return return_map_pair_string(); },
        py::return_value_policy_pack({{rvpb, rvpc}, {rvpc, rvpb}}));

    m.def(
        "return_set_sb",
        []() { return return_set_pair_string(); },
        py::return_value_policy_pack({rvpc, rvpb}));
    m.def(
        "return_set_bs",
        []() { return return_set_pair_string(); },
        py::return_value_policy_pack({rvpb, rvpc}));

    m.def(
        "return_vector_sb",
        []() { return return_vector_pair_string(); },
        py::return_value_policy_pack({rvpc, rvpb}));
    m.def(
        "return_vector_bs",
        []() { return return_vector_pair_string(); },
        py::return_value_policy_pack({rvpb, rvpc}));

    m.def(
        "return_array_sb",
        []() { return return_array_pair_string(); },
        py::return_value_policy_pack({rvpc, rvpb}));
    m.def(
        "return_array_bs",
        []() { return return_array_pair_string(); },
        py::return_value_policy_pack({rvpb, rvpc}));

    m.def(
        "return_optional_sb",
        []() { return return_optional_pair_string(); },
        py::return_value_policy_pack({rvpc, rvpb}));
    m.def(
        "return_optional_bs",
        []() { return return_optional_pair_string(); },
        py::return_value_policy_pack({rvpb, rvpc}));

    m.def(
        "return_variant_sb",
        []() { return return_variant_pair_string(); },
        py::return_value_policy_pack({rvpc, rvpb}));
    m.def(
        "return_variant_bs",
        []() { return return_variant_pair_string(); },
        py::return_value_policy_pack({rvpb, rvpc}));

    // Here the rvp is applied to the return value of call_callback_pass_pair_string:
    m.def("call_callback_pass_pair_string_rtn_s", call_callback_pass_pair_string);
    m.def("call_callback_pass_pair_string_rtn_b", call_callback_pass_pair_string, rvpb);
}

File: tests/test_return_value_policy_pack.py
import pytest

from pybind11_tests import return_value_policy_pack as m


@pytest.mark.parametrize(
    ("func", "expected"),
    [
        (m.return_tuple_str_str, (str, str)),
        (m.return_tuple_bytes_bytes, (bytes, bytes)),
        (m.return_tuple_str_bytes, (str, bytes)),
        (m.return_tuple_bytes_str, (bytes, str)),
    ],
)
def test_return_pair_string(func, expected):
    p = func()
    assert isinstance(p, tuple)
    assert len(p) == 2
    assert all(isinstance(e, t) for e, t in zip(p, expected))


@pytest.mark.parametrize(
    ("func", "expected"),
    [
        (m.return_nested_tuple_str, (str, str, str, str)),
        (m.return_nested_tuple_bytes, (bytes, bytes, bytes, bytes)),
        (m.return_nested_tuple_str_bytes_bytes_str, (str, bytes, bytes, str)),
        (m.return_nested_tuple_bytes_str_str_bytes, (bytes, str, str, bytes)),
    ],
)
def test_return_nested_pair_string(func, expected):
    np = func()
    assert isinstance(np, tuple)
    assert len(np) == 2
    assert all(isinstance(e, t) for e, t in zip(sum(np, ()), expected))


@pytest.mark.parametrize(
    ("func", "expected"),
    [
        (m.return_dict_str_str, (str, str)),
        (m.return_dict_bytes_bytes, (bytes, bytes)),
        (m.return_dict_str_bytes, (str, bytes)),
        (m.return_dict_bytes_str, (bytes, str)),
    ],
)
def test_return_map_string(func, expected):
    mp = func()
    assert isinstance(mp, dict)
    assert len(mp) == 1
    assert all(isinstance(e, t) for e, t in zip(tuple(mp.items())[0], expected))


@pytest.mark.parametrize(
    ("func", "expected"),
    [
        (m.return_map_sbbs, (str, bytes, bytes, str)),
        (m.return_map_bssb, (bytes, str, str, bytes)),
    ],
)
def test_return_dict_pair_string(func, expected):
    mp = func()
    assert isinstance(mp, dict)
    assert len(mp) == 1
    assert all(
        isinstance(e, t) for e, t in zip(sum(tuple(mp.items())[0], ()), expected)
    )


@pytest.mark.parametrize(
    ("func", "expected"),
    [
        (m.return_set_sb, (str, bytes)),
        (m.return_set_bs, (bytes, str)),
    ],
)
def test_return_set_pair_string(func, expected):
    sp = func()
    assert isinstance(sp, set)
    assert len(sp) == 1
    assert all(isinstance(e, t) for e, t in zip(sum(tuple(sp), ()), expected))


@pytest.mark.parametrize(
    ("func", "expected"),
    [
        (m.return_vector_sb, (str, bytes)),
        (m.return_vector_bs, (bytes, str)),
        (m.return_array_sb, (str, bytes)),
        (m.return_array_bs, (bytes, str)),
    ],
)
def test_return_list_pair_string(func, expected):
    vp = func()
    assert isinstance(vp, list)
    assert len(vp) == 1
    assert all(isinstance(e, t) for e, t in zip(sum(vp, ()), expected))


def _test_return_optional_variant_pair_string(fname, expected):
    func = getattr(m, fname)
    p = func()
    assert isinstance(p, tuple)
    assert len(p) == 2
    assert all(isinstance(e, t) for e, t in zip(p, expected))


@pytest.mark.skipif(not m.PYBIND11_HAS_OPTIONAL, reason="<optional> not available.")
@pytest.mark.parametrize(
    ("fname", "expected"),
    [
        ("return_optional_sb", (str, bytes)),
        ("return_optional_bs", (bytes, str)),
    ],
)
def test_return_optional_pair_string(fname, expected):
    _test_return_optional_variant_pair_string(fname, expected)


@pytest.mark.skipif(not m.PYBIND11_HAS_VARIANT, reason="<variant> not available.")
@pytest.mark.parametrize(
    ("fname", "expected"),
    [
        ("return_variant_sb", (str, bytes)),
        ("return_variant_bs", (bytes, str)),
    ],
)
def test_return_variant_pair_string(fname, expected):
    _test_return_optional_variant_pair_string(fname, expected)


@pytest.mark.parametrize(
    ("func", "expected"),
    [
        (m.call_callback_pass_pair_default, repr(("", ""))),
        (m.call_callback_pass_pair_s, repr(("", ""))),
        (m.call_callback_pass_pair_b, repr((b"", b""))),
        (m.call_callback_pass_pair_sb, repr(("", b""))),
        (m.call_callback_pass_pair_bs, repr((b"", ""))),
    ],
)
def test_call_callback_pass_pair_string(func, expected):
    def cb(p):
        assert isinstance(p, tuple)
        assert len(p) == 2
        return repr(p)

    assert func(cb) == expected


def test_nested_callbacks_si_s():
    def cb_1(i):
        assert isinstance(i, int)
        return "cb_1_" + str(i)

    def cb_2(cb):
        r = cb(20)
        assert isinstance(r, str)
        return "cb_2_" + r

    def cb_3(cb):
        r = cb(cb_1)
        assert isinstance(r, str)
        return "cb_3_" + r

    def cb_4(cb):
        r = cb(cb_2)
        assert isinstance(r, str)
        return "cb_4_" + r

    assert m.call_level_1_callback_si_s(cb_1) == "cb_1_1001"
    assert m.call_level_2_callback_si_s(cb_2) == "cb_2_level_0_si_20"
    assert m.call_level_3_callback_si_s(cb_3) == "cb_3_cb_1_1001"
    assert m.call_level_4_callback_si_s(cb_4) == "cb_4_cb_2_level_0_si_20"


def test_nested_callbacks_si_b():
    def cb_1(i):
        assert isinstance(i, int)
        return b"\x80cb_1_" + (str(i)).encode()

    def cb_2(cb):
        r = cb(20)
        assert isinstance(r, bytes)
        return b"\x80cb_2_" + r

    def cb_3(cb):
        r = cb(cb_1)
        assert isinstance(r, bytes)
        return b"\x80cb_3_" + r

    def cb_2s(cb):
        r = cb(20)
        assert isinstance(r, str)  # Passing bytes is unsupported (missing feature).
        return b"\x80cb_2_" + r.encode()

    def cb_4(cb):
        r = cb(cb_2s)
        assert isinstance(r, bytes)
        return b"\x80cb_4_" + r

    assert m.call_level_1_callback_si_b(cb_1) == b"\x80cb_1_1001"
    assert m.call_level_2_callback_si_b(cb_2) == b"\x80cb_2_level_0_si_20"
    assert m.call_level_3_callback_si_b(cb_3) == b"\x80cb_3_\x80cb_1_1001"
    assert m.call_level_4_callback_si_b(cb_4) == b"\x80cb_4_\x80cb_2_level_0_si_20"


def test_nested_callbacks_is_s():
    def cb_1(s):
        assert isinstance(s, str)
        return 10000 + int(s)

    def cb_2(cb):
        return 20000 + cb("20")

    def cb_3(cb):
        return 30000 + cb(cb_1)

    def cb_4(cb):
        return 40000 + cb(cb_2)

    assert m.call_level_1_callback_is_s(cb_1) == 10101
    assert m.call_level_2_callback_is_s(cb_2) == 20120
    assert m.call_level_3_callback_is_s(cb_3) == 40101
    assert m.call_level_4_callback_is_s(cb_4) == 60120


def test_nested_callbacks_is_b():
    def cb_1(s):
        assert isinstance(s, bytes)
        return 10000 + int(s)

    def cb_2(cb):
        return 20000 + cb(b"20")

    def cb_1s(s):
        assert isinstance(s, str)  # Passing bytes is unsupported (missing feature).
        return 10000 + int(s)

    def cb_3(cb):
        return 30000 + cb(cb_1s)

    def cb_4(cb):
        return 40000 + cb(cb_2)

    assert m.call_level_1_callback_is_b(cb_1) == 10101
    assert m.call_level_2_callback_is_b(cb_2) == 20120
    assert m.call_level_3_callback_is_b(cb_3) == 40101
    assert m.call_level_4_callback_is_b(cb_4) == 60120
